main_IK.py

```diff
@@ -0,0 +1,160 @@
+from Gauss_Newton_IK import GaussNewtonIK
+import mujoco
+import mujoco.viewer
+import numpy as np
+import matplotlib
+matplotlib.use('Agg')  # Use non-GUI backend
+import matplotlib.pyplot as plt
+
+# Load model and data
+model = mujoco.MjModel.from_xml_path("scene.xml")
+data = mujoco.MjData(model)
+
+# PI control gains
+Kp = np.array([100, 10, 10, 50, 50, 50])  
+Ki = np.array([0.5, 0.5, 0.5, 0.1, 0.1, 0.1]) 
+
+num_joints = model.nu
+
+# Sinusoidal trajectory parameters
+duration = 10.0  # simulation time in seconds
+frequency = 0.02  # Hz
+amplitude = np.deg2rad(30)  # ±30 degrees
+
+# Trajectory history buffers
+q_des_history = []
+qd_des_history = []
+q_act_history = []
+qd_act_history = []
+tau_history = []
+time_history = []
+
+# Initial pose
+initial_qpos = np.copy(data.qpos[:num_joints])
+q_des = np.zeros(num_joints)
+
+# Initialize IK solver
+ik_solver = GaussNewtonIK(model, data)
+
+# Integral term
+integral_error = np.zeros(num_joints)
+sim_start = data.time
+base_target = np.array([0.25, 0.5, 0.5])
+
+# Sinusoid tracking 
+amp_x = 0.1 
+amp_y = 0.1 
+amp_z = 0.1 
+freq_x = 0.1
+freq_y = 0.2 
+freq_z = 0.1
+
+body_name = "wrist_3_link"
+body_id = model.body(body_name).id
+jacp = np.zeros((3, model.nv))
+jacr = np.zeros((3, model.nv))
+
+with mujoco.viewer.launch_passive(model, data) as viewer:
+    for _ in range(int(duration / model.opt.timestep)):
+        t = data.time - sim_start
+        
+        J_pos = np.zeros(3 * model.nv)
+        J_rot = np.zeros(3 * model.nv)
+
+        target_pos = base_target + np.array([amp_x * np.sin(2 * np.pi * freq_x * t),amp_y * np.sin(2 * np.pi * freq_y * t),amp_z * np.sin(2 * np.pi * freq_z * t)])
+        
+        q_des = ik_solver.solve(target_pos, body_name="wrist_3_link", q_init=data.qpos[:model.nq])
+        
+        x_vel = np.array([amp_x*(2*np.pi*freq_x)*np.cos(2 * np.pi * freq_x * t), amp_y*(2*np.pi*freq_y)*np.cos(2 * np.pi * freq_y * t), amp_z*(2*np.pi*freq_z)*np.cos(2 * np.pi * freq_z * t)])
+        x_acc = np.array([-amp_x *(2*np.pi*freq_x)**2*np.sin(2 * np.pi * freq_x * t),-amp_y*(2*np.pi*freq_y)**2 * np.sin(2 * np.pi * freq_y * t), -amp_z*(2*np.pi*freq_z)**2 * np.sin(2 * np.pi * freq_z * t)])
+        
+        x_current = data.body(body_id).xpos.copy()
+        
+        mujoco.mj_jac(model, data, jacp, jacr, x_current, body_id)  
+        J = jacp[:, :model.nv]
+        
+        qd_des = np.linalg.pinv(J) @ x_vel  
+        qdd_des = np.linalg.pinv(J) @ x_acc
```

Comment:
This pseudoinverse as-is will have issues near singularities, which is why the "damped least squares" or "Levenberg-Marquardt" version exists, which adds a small scale identity matrix to the Jacobian.

Compare sections 4 and 5 here, where you have implemented section 4: https://www.cs.cmu.edu/~15464-s13/lectures/lecture6/iksurvey.pdf

Also worth making a separate issue / PR for this!